### DISCUSSION TITLE
Programmatically restore a DB in the cloud

### DISCUSSION THREAD

**Discussion by ssteiner** [ravendb/ravendb#19046] [Q&A]
I'm looking at the SDK documentation for the `RestoreBackupOperation` [here](https://ravendb.net/docs/article-page/6.0/csharp/client-api/operations/maintenance/backup/restore#restoring-a-database:-configuration-and-execution). Looking at the properties of the `RestoreBackupConfiguration`, it looks rather straightforward - that is for an on-prem deployment. We have the name of the database, and the pack, plus `LastFileNameToRestore` for incremental backups.

But, my RavenDb instance is in the cloud - how does it work there? `DatabaseName` is the only thing that's clear to me. I'm on the automatic backup train, so I don't even have access to where the backups are being stored.

**Comment by ayende** [comment]
You need to use `RestoreFromS3Configuration`
In the Cloud Portal, you can get the backup restore URL, if you'll access that URL, you'll have the S3 settings for the configuration

**Comment by ssteiner** [comment]
I take it you mean follow [these instruction](https://ravendb.net/docs/article-page/6.0/csharp/cloud/cloud-backup-and-restore#restore-mandatory-backup-files)s to get the restore link, correct? I'm on the free tier and my backup page looks like this:

![image](https://github.com/user-attachments/assets/dbe7f01c-9d93-4c51-aaf8-a06a6d49078e)

am I missing something?

**Comment by gregolsky** [comment]
@ssteiner To get a restore link you need to visit Backups section in the RavenDB Cloud portal after logging in at https://cloud.ravendb.net/user/login

![image](https://github.com/user-attachments/assets/bd9c1e62-fd34-4f5c-aaa8-da5c76a2fe66)

**Reply by ssteiner** [reply]
Okay, we're almost there. One thing I'm unsure about is` RestoreFromS3Configuration.EncryptionKey` versus `RestoreFromS3Configuration.BackupEncryptionSettings.Key`. There seems to be a restore key for non encrypted databases, and a database encryption key. Which one goes where?

I then tried to run the whole thing using

```
RestoreFromS3Configuration restoreConfig = new()
{
    DatabaseName = databaseName,  
    Settings = new S3Settings { CustomServerUrl = backupKey }, 
    //EncryptionKey = encryptionKey
};
if (!string.IsNullOrEmpty(encryptionKey))
    restoreConfig.BackupEncryptionSettings = new BackupEncryptionSettings { EncryptionMode = EncryptionMode.UseDatabaseKey, Key = encryptionKey };

var restoreBackupTask = new RestoreBackupOperation(restoreConfig);
var op = Store.Maintenance.Server.Send(restoreBackupTask);
var opRes = await op.WaitForCompletionAsync().ConfigureAwait(false);
```

which bombed out with a message

> System.ArgumentException: AWS Access Key cannot be null or empty

So where do I put the backup link I get from the studio to make this work?

Also, is there a way to list the backups and get the backup link programmatically?

**Reply by grisha-kotler** [reply]
Hi @ssteiner,

You need to use the restore link in order to get the AWS credentials.
After that you need to fill the `S3Settings` with the output from the previous step.

**Reply by ssteiner** [reply]
Is there any API to process the response from the backup link, or would I just have to write my own? It seems the response from the backup URL matches the `S3Settings` class pretty well.

I found a problem with the UI in Studio. Look at point 7 -

> If your database is encrypted please 

and then there should be some kind of verb.

![image](https://github.com/user-attachments/assets/ca2a01f2-3c4f-4135-a803-b214380679ea)

With the S3Settings from the backup link, I finally managed to properly restore after a lot of trial & error. So it seems the restoreConfig needs to look like this:

```
RestoreFromS3Configuration restoreConfig = new()
{
    DatabaseName = databaseName,  
    Settings = awsSettings,
    BackupEncryptionSettings = new BackupEncryptionSettings { EncryptionMode = EncryptionMode.UseProvidedKey, Key = encryptionKey}, 
};
```

where `encryptionKey` is the restore key from the studio

Restore would then work like this:

```
var restoreBackupTask = new RestoreBackupOperation(restoreConfig);
var op = Store.Maintenance.Server.Send(restoreBackupTask);
var opRes = await op.WaitForCompletionAsync().ConfigureAwait(false);
```

What would I look for in `opRes` to determine that all is well? Can I assume no exception => all okay or is there something to look out for?

Also, in studio, I see the individual backups.. is there a way to list those using an API and generate a backup link programmatically?

![image](https://github.com/user-attachments/assets/08c01067-a72d-4da3-a11b-d0d473a5d229)

**Reply by grisha-kotler** [reply]
- Regarding the UI, we'll correct the phrasing.

- `var opRes = await op.WaitForCompletionAsync().ConfigureAwait(false);`
If the operation fails, it will throw an exception.

- We do not have a public API available for generating the backup link.
